cmd/kube-proxy/app/server.go

```diff
@@ -1046,7 +1046,8 @@ func getNodeIPs(logger klog.Logger, client clientset.Interface, name string) []n
 	}
 
 	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
-		node, err := client.CoreV1().Nodes().Get(context.TODO(), name, metav1.GetOptions{})
+		// Improve the efficiency by setting ResourceVersion "0", which means "serve the current version held by the API server cache".
+		node, err := client.CoreV1().Nodes().Get(context.TODO(), name, metav1.GetOptions{ResourceVersion: "0"})
```

Comment:
@aojea @danwinship - can the HostIPs that we get couple lines below can change (hopefully not) or can be added in steps (e.g. IPv4 address is added first, and in a subsequent call IPv6 - or literally any other scenario).

Given that it's an initialization phase (and one-time thing), if we get the wrong one, it may be problemtic.

[OTOH, we don't coordinate start of kube-proxy vs other components, e.g. kubelet, so if that was a problem, we would already be exposed to it, so I guess this change is probably ok...]

Reply:
usually is the same controller doing the address population, but we had a bug where the address changed after the start https://github.com/kubernetes/kubernetes/pull/121028 , it is fixed now but node addresses are super tricky, I think we should be conservative here, it is not really much of an opimization if we open this new failures modes

Reply:
I feel the bug is unrelated to stale data but `kube-proxy` being unable to reconcile with data changes after startup. 

Hitting the etcd or not, `kube-proxy` might still read the data, which is changed right after

> I think we should be conservative here, it is not really much of an opimization if we open this new failures modes

Agreed that we should be conservative here but the optimization is noticeable as mentioned [here](https://github.com/kubernetes/kubernetes/pull/123286#discussion_r1491394175)

Reply:
fwiw, the informer, which is used everywhere in the kube-proxy, starts with "ResourceVersion: 0" as well. See [code](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L844)

Reply:
I asked Eric to hold on this, there were several problems with spot instances  and kube-proxy related to stale data that consumed a lot of my time just to figure it out and this may or may not have this risk ..., as soon as I can get some more time I will go with him through all the possible scenarios to be sure this chane is safe, this was this way for a long time so I don't think we are in a rush now

Reply:
I actually found a potential but seemingly harmless "bug" here. 

In my env, `kube-proxy` also runs standalone so at the startup the node is not created until a few minutes later. However, we [specified retry steps](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-proxy/app/server.go#L1042) with 6 and give up on getting NodeIPs. In other words, the NodeIPs are never retrieved from apiserver and assumed with loopback IP (see logs below).

`Can't determine this node's IP, assuming loopback; if this is incorrect, please set the --bind-address flag`

So IMHO the reconciling on NodeIPs may be necessary or we should fix the getNodeIPs() to match the comment "If required, it will wait for the node to be created."